### ISSUE TITLE
Using multi-config packages with CMake

### ISSUE THREAD

**Issue by slnovak** [conan-io/conan#2983]
Hi,

I'm using Conan v1.3.2. I'm currently trying to use upstream multi-config packages that specify `cppinfo.{debug|release}.{...}` for using debug and release libraries within a downstream package and corresponding `build_type`. I'm a bit confused as to how to use these dependencies while trying to develop and build the downstream package. I have a few specific questions:

1. What are the consequences of specifying `cppinfo.X`, `cppinfo.debug.X`, _and_ `cppinfo.release.X` in a package? I see that in a downstream package that uses the CMake generator, each of these values are interpolated in `conanbuildinfo.cmake`, specifically with the `set(CONAN_X_ROOT ...)`, `set(CONAN_X_ROOT_RELEASE ...)`, and `set(CONAN_X_ROOT_DEBUG ...)` commands. Can you clarify how each are used with the (single-configuration) CMake generator?

2. When running `conan install .. -s build_type=Debug` (in `$PROJECT_ROOT/build`) followed by `conan build . -bf build` (in `$PROJECT_ROOT`), I expect that my CMake configuration would configure the resulting Visual Studio solution to link with the debug libraries in each of the upstream packages as opposed to Release libraries). However, it appears that Conan is linking to the _release_ libraries. How can I configure the CMake configuration to link to the corresponding library specified by `build_type`?

Thanks for your help. I've sifted through the docs a handful of times but have trouble trying to understand what the conventions/limitations are for using the CMake generator across different build types.

Cheers!

**Comment by lasote** [comment]
Hi!
I think in both 1 and 2, if you specify a regular `cmake` generator and you are using upstream multiconfiguration packages you can better use `conan_target_link_libraries` macro, example [here](https://github.com/memsharded/hello_multi_config/blob/master/test_package/CMakeLists.txt) that will link the libraries depending on the specific `build_type`. Let me know if it helps:

```
macro(conan_target_link_libraries target)
    if(CONAN_TARGETS)
        target_link_libraries(${target} ${CONAN_TARGETS})
    else()
        target_link_libraries(${target} ${CONAN_LIBS})
        foreach(_LIB ${CONAN_LIBS_RELEASE})
            target_link_libraries(${target} optimized ${_LIB})
        endforeach()
        foreach(_LIB ${CONAN_LIBS_DEBUG})
            target_link_libraries(${target} debug ${_LIB})
        endforeach()
    endif()
endmacro()
```

**Comment by lasote** [comment]
By the way, probably this is related with: https://github.com/conan-io/conan/issues/2981
To be solved asap in the 1.4.2

**Comment by memsharded** [comment]
> What are the consequences of specifying cppinfo.X, cppinfo.debug.X, and cppinfo.release.X in a package? I see that in a downstream package that uses the CMake generator, each of these values are interpolated in conanbuildinfo.cmake, specifically with the set(CONAN_X_ROOT ...), set(CONAN_X_ROOT_RELEASE ...), and set(CONAN_X_ROOT_DEBUG ...) commands. Can you clarify how each are used with the (single-configuration) CMake generator?

The single configuration ``cmake`` can use the debug/release information transparently if using TARGETS approach, but if using the global cmake approach, you must be aware, and use the ``conan_target_link_libraries`` macro or handle yourself CONAN_LIBS_RELEASE/DEBUG variables.

In your point 2), yes, it could be related to that issue, please upgrade to 1.4.2 when released and report if still failing. Thanks!

**Comment by slnovak** [comment]
Great, thank you for the quick response!

**Comment by slnovak** [comment]
Hi! Thanks again for your help. It looks like my issue was resolved with 1.4.2. Cheers! :beers:

**Comment by memsharded** [comment]
Great! Thanks for reporting and closing!
